main.py

```diff
@@ -1145,37 +1148,45 @@ def update_file_diagnostics(window: sublime.Window, file_path: str, source: str,
 phantom_sets_by_buffer = {}  # type: Dict[int, sublime.PhantomSet]
 
 
-def update_diagnostics_in_view(view: sublime.View, diagnostics: 'List[Diagnostic]'):
+def update_diagnostics_phantoms(view: sublime.View, diagnostics: 'List[Diagnostic]'):
     global phantom_sets_by_buffer
 
-    phantoms = []  # type: List[sublime.Phantom]
-    regions = []  # type: List[sublime.Region]
+    buffer_id = view.buffer_id()
+    if show_diagnostics_phantoms and not view.is_dirty():
+        phantoms = list(
+            create_phantom(view, diagnostic) for diagnostic in diagnostics)
+    else:
+        phantoms = None  # type: ignore
+    if phantoms:
+        phantom_set = phantom_sets_by_buffer.get(buffer_id)
+        if not phantom_set:
+            phantom_set = sublime.PhantomSet(view, "lsp_diagnostics")
+            phantom_sets_by_buffer[buffer_id] = phantom_set
+        phantom_set.update(phantoms)
+    else:
+        phantom_sets_by_buffer.pop(buffer_id, None)
 
-    if view is not None:
-        if view.is_dirty() or not show_diagnostics_phantoms:
-            regions = list(
-                create_region(view, diagnostic) for diagnostic in diagnostics)
-        else:
-            phantoms = list(
-                create_phantom(view, diagnostic) for diagnostic in diagnostics)
 
-        # TODO: if phantoms are disabled, this logic can be skipped
-        buffer_id = view.buffer_id()
-        if buffer_id not in phantom_sets_by_buffer:
-            phantom_set = sublime.PhantomSet(view, "diagnostics")
-            phantom_sets_by_buffer[buffer_id] = phantom_set
-        else:
-            phantom_set = phantom_sets_by_buffer[buffer_id]
+def update_diagnostics_regions(view: sublime.View, diagnostics: 'List[Diagnostic]', severity: int):
+    region_name = "lsp_" + format_severity(severity)
+    if show_diagnostics_phantoms and not view.is_dirty():
+        regions = None  # type: ignore
```

Comment:
Thanks for giving this a go, mypy-pyls also did not report this issue, running `mypy main.py` locally did.
I agree the type inference is not 100%, but I'm glad it could be solved without extra annotations.